### PR TITLE
Unicode sys tests and fixes

### LINKED ISSUE #8094: Sys unicode unit tests
I know it's late since we're already at RC2 but I'm a bit worried about our state of Sys support wrt unicode, as #8087 #8057 etc are showing.

I don't want to delay the 4.0 final release but I think it's better if we have a good picture of were we're at and if we can fix some of the most urgent issues before final release that would be great.

We need to test the following (with full unicode chars - including 😂):
- filesystem names (file/dirs) , cwd/setcwd
- filesystem I/O (read/write)
- Sys.args
- print
- stdin/stderr/stdout
- sys.io.Process args + i/o
- env variables get/put
- and basically most of Sys class API

The tricky part is that we cannot just check that `readDirectory(createDirectory("é")) == "é"` since we might create some weirdly name directory, then decode it wrongly, and consider that we pass the test.

So for instance we need to have first a validation of readDirectory (with an already GIT created dir+file) and then we can test createDirectory.

Same of env vars / Sys.args etc.

### PR DESCRIPTION
**Pull request by Aurel300**
Closes #8094.

## General ##

 - ~~define `target.unicode` - true by default, false on Neko or with `(cpp && !cppia && !hxcpp_smart_strings)`~~ (extracted into a separate PR #8244)

## Test setup ##

 - [x] create a script to generate a test suite with Unicode-named files and directories
   - no NUL bytes
   - no /\:*"?<>|
   - UTF-8 boundary conditions
   - non-BMP characters
   - NFC/NFD ambiguous characters
   - ~~compatibility equivalent characters?~~ (out of scope)

## Filesystem ##

 - `Sys`
   - [x] `setCwd`, `getCwd` (cd into Uni directory, read back path)
   - ~~`programPath` (symlink in a Uni directory, read back path)~~ this does not work over symlinks – we could copy binaries into Uni directories and run them there, but this is problematic to cover for all targets; `getCwd` should mostly cover this anyway
 - `sys.FileSystem`
   - [x] `absolutePath` (after `setCwd`, nested)
   - [x] `exists`
   - [x] `fullPath` (with symlinks in Uni directories)
   - [x] `isDirectory`
   - [x] `readDirectory` (test dir + nested Uni directories)
   - [x] `stat`
   - [x] `rename`
   - [x] `deleteFile`
   - [x] `deleteDirectory`
   - [x] `createDirectory`
 - `sys.io.File`
   - [x] `copy` (this should probably be in `sys.FileSystem`)
 - `haxe.io.Path`
   - [x] construct + check directory components, name components, extension components

## I/O API ##

Read from a known source. Write to a temporary file and compare on `Bytes` against a known correct output.

 - [x] `trace`
 - `sys.io.File`
   - [x] `getBytes`
   - [x] `getContent`
   - [x] `read` (+ `haxe.io.Input` test)
   - [x] `saveContent`
   - [x] `update`, `write`, `append` (+ `haxe.io.Output` test)
   - [x] Unicode *paths* for the above
 - `haxe.io.Input`
   - [x] `readLine`
   - [x] `readString`
   - [x] `readUntil`
 - `haxe.io.Output`
   - [x] `writeString`
 - `Sys`
   - [x] `stdin`, `stderr`, `stdout`
   - [x] `print`, `println`
   - ~~`getChar`~~ (not Unicode related)

## Environment-related API ##

 - `Sys`
   - [x] `args`
   - [x] `environment`
   - [x] `getEnv`, `putEnv`

## Process API ##

 - `Sys`
   - [ ] `command` - Uni executable, Uni arguments
 - `sys.io.Process`
   - [x] `new`
   - [x] `stdin`, `stderr`, `stdout`

## Miscellaneous ##

 - [x] move sys Unicode tests into `TestUnicode`
 - [x] ~~`genTestRes.sh` Windows port~~ ~~commit test suite to git~~ just use Python
 - ~~some common scaffolding for running process of the same platform (currently in `TestUnicode` as well as `TestCommandBase`)~~
 - ~~replace `ExitCode` with `UtilityProcess`~~

---

## Check targets ##

 - Travis
   - [x] cpp
   - [x] cs
   - [x] eval
   - [x] hl
   - [x] lua
   - [x] java
   - [x] jvm
   - [x] php
   - [x] python
 - AppVeyor
   - [x] cpp
   - [x] cs
   - [x] eval
   - [x] hl
   - [x] java
   - [x] jvm
   - [x] php
   - [x] python

## Issues / PRs spawned ##

 - [x] (C++) `trace`, `Sys.print`, `Sys.println`, `String.split`, UTF-8 decoding OOB https://github.com/HaxeFoundation/hxcpp/pull/781
 - [x] (Neko / HL) `Utf8.iter` (in native code) throws for `\x7F` https://github.com/HaxeFoundation/haxe/issues/8139
 - [x] ~~`Input.readString` should read read N _characters_~~ #8199 (bytes it is!)
 - [x] (PHP) `StringTools.endsWith` fails on Unicode strings #8211
 - [x] (PHP) `Bytes.getString` with non-BMP sequences #8212
 - [x] (HL) `haxe.io.Path` uses `StringTools.fastCodeAt` without handling surrogates (fix is part of this PR)
 - [x] (C++) `Sys.args` does not decode UTF-8 arguments correctly https://github.com/HaxeFoundation/hxcpp/pull/783
 - [x] (C++) UTF-16 surrogates are incorrectly detected https://github.com/HaxeFoundation/hxcpp/pull/782
 - [x] (HL) Surrogate issues https://github.com/HaxeFoundation/hashlink/pull/246
 - [x] (HL) U+10000 swallows next character https://github.com/HaxeFoundation/haxe/issues/8245
 - [x] (Python) `Output.writeString` writes bytes individually as codepoints #8197
 - [x] (Python) `Sys.stdin().readLine()` fails on Unicode input #8374 
 - [x] (Python, PHP, eval) `Sys.println()` should append `\r\n` on Windows (fix is part of this PR)
 - [x] (Python) `trace` defaults to `cp1252` on Windows when piped (fix is part of this PR)
 - [x] (C#, Java, JVM) `Sys.stdout` may not default to UTF-8 on Windows (fix is part of this PR)
 - [ ] (C++) UTF-16 decoding OOB
 - [ ] (C#, Lua) `FileSystem.fullPath` does not resolve symlinks #8215
 - [ ] (C#) `FileSystem.readDirectory` reports bad encoding #8247 
 - [ ] (Lua) `Sys.environment` segfault #8216
 - [ ] (C++, HL) `Sys.print` duplicates CR on Windows #8379
 - [ ] (C++) `Sys.setCwd`, `.deleteFile`, `.deleteDirectory`, `FileSystem.exists`, `.rename` fail on Unicode path #8400

### CODE REVIEW

**Comment by RealyUniqueName** [comment]
`sys.io.File` is missing on your list.

**Comment by RealyUniqueName** [comment]
Also, a windows script for unicode filenames generation is needed.

**Comment by Aurel300** [comment]
> `sys.io.File` is missing on your list.

Yup, I only listed the methods for `FileSystem`, didn't go through `File` in detail yet.

> Also, a windows script for unicode filenames generation is needed.

I might port it yet, although the idea was partly to generate the test suite and commit the resulting files and directories to git.

File: src/context/common.ml [hunk 1]
```diff
@@ -283,6 +285,7 @@ let default_config =
 		pf_uses_utf16 = true;
 		pf_this_before_super = true;
 		pf_supports_threads = false;
+		pf_supports_unicode = false;
```
Comment (RealyUniqueName) [inline src/context/common.ml]: true?

Comment (Aurel300) [reply src/context/common.ml]: Yeah, my bad.

**Comment by ncannasse** [comment]
Given there's not much fixes to do I'm putting that in Release 4.0 to monitor the changes. In case we have some very target specific issue yet unresolved we will still not block final release because of this.

**Comment by Aurel300** [comment]
So the non-Windows part of this is basically done. The various target failures should be covered by the "Issues / PRs spawned" list. I am not sure how to proceed with Windows support here (cc @andyli):

The last commit attempted to simply commit the test suite directory with its Unicode-named files and sub-directories. But AppVeyor fails to even `git clone` the repository with this. Apparently `chcp 65001` should be run beforehand, but `git clone` is not even in the CI file, it is run automatically.

I also tried to put the `test-res` directory into a ZIP archive. Unfortunately this seems to fail on OS X already (it tries to override existing files, because `\x01` is completely ignored, I think?) … Apparently this is best "solved" by using a different program, e.g. `7z`.

Porting `genTestRes.sh` would be best, but outputting the correct Unicode filenames seems to be rather painful in Windows batch. I found workarounds to simply put binary data into a file (e.g. https://stackoverflow.com/questions/47750732/write-hex-values-to-file-in-windows-batch), but nothing yet to actually create files and directories with the correct names.

**Comment by andyli** [comment]
7z is pre-installed in AppVeyor: https://www.appveyor.com/docs/windows-images-software/#tools

You can also use whatever program that can correctly handle Unicode file names correctly (e.g. Python?).

One tip about working with AppVeyor: You can [remote desktop](https://www.appveyor.com/docs/how-to/rdp-to-build-worker/) into the build and do whatever testing/debugging there. Just make sure you're not doing it in the HF repo, but use a fork of your own, to prevent other people connect into it and print out our secret variables (those wouldn't be available in fork builds).

File: std/haxe/io/Path.hx [hunk 2]
```diff
@@ -230,8 +230,13 @@ class Path {
 		var acc = new StringBuf();
 		var colon = false;
 		var slashes = false;
+		#if hl
```
Comment (ncannasse) [inline std/haxe/io/Path.hx]: should it not be #if utf16 ?

**Comment by ncannasse** [comment]
I have merged the HL PR and restarted the build to update results.

File: std/haxe/io/Path.hx (hunk 2 above)
Comment (Aurel300) [reply std/haxe/io/Path.hx]: It was (in bba3863), but that broke the Java tests.

**Comment by ncannasse** [comment]
Uhm, something is weird there then, how and why were the tests were broken?

**Comment by Aurel300** [comment]
> Uhm, something is weird there then, how and why were the tests were broken?

Here are the failing tests: https://travis-ci.org/HaxeFoundation/haxe/jobs/524618714#L3553

I didn't investigate why they failed, but you are right, it really should work on Java. Interestingly enough, the failures are e.g. expected `0x1F619`, but got `0xF619` (forgetting to add the `0x10000`), and this is through the Unicode-safe string iterator, I think.

I'll investigate if there is a problem with the iterator on Java tomorrow. I'm also wondering why it even works as-is on Java.

**Comment by Aurel300** [comment]
Ok, fixed that with #8239.

File: std/haxe/io/Path.hx (hunk 2 above)
Comment (ncannasse) [reply std/haxe/io/Path.hx]: It would be interesting to know why this break the Java tests... And what about JS ? it's also utf16.

Comment (Aurel300) [reply std/haxe/io/Path.hx]: #8239 fixed this. I changed this back to `#if utf16` just now. JS is actually works with both versions.

**Comment by Aurel300** [comment]
It would be best if somebody with a working Haxe setup on Windows looked into this. It is difficult for me to debug the targets from AppVeyor results.

---

Update: I got a Vagrant set up with Windows 10 that mostly works, so hopefully I can progress on this again.

**Comment by Aurel300** [comment]
Travis failure is unrelated (`brew` problem when installing Flash player).
